pkg/exchange/bitget/stream.go

```diff
@@ -0,0 +1,167 @@
+package bitget
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/c9s/bbgo/pkg/exchange/bitget/bitgetapi"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+//go:generate callbackgen -type Stream
+type Stream struct {
+	types.StandardStream
+
+	booksEventCallbacks []func(o BookEvent)
+}
+
+func NewStream() *Stream {
+	stream := &Stream{
+		StandardStream: types.NewStandardStream(),
+	}
+
+	stream.SetEndpointCreator(stream.createEndpoint)
+	stream.SetParser(stream.parseWebSocketEvent)
+	stream.SetDispatcher(stream.dispatchEvent)
+	stream.OnConnect(stream.handlerConnect)
+
+	stream.OnBooksEvent(stream.handleBooksEvent)
+	return stream
+}
+
+func (s *Stream) syncSubscriptions(opType WsEventType) error {
+	if opType != WsEventUnsubscribe && opType != WsEventSubscribe {
+		return fmt.Errorf("unexpected subscription type: %v", opType)
+	}
+
+	logger := log.WithField("opType", opType)
+	args := []WebsocketArg{}
+	for _, subscription := range s.Subscriptions {
+		arg, err := convertSubscription(subscription)
+		if err != nil {
+			logger.WithError(err).Errorf("convert error, subscription: %+v", subscription)
```

Comment:
can simply return the error here I think

Reply:
oh, it cannot be. The reason is that the `syncSubscriptions` function is invoked by both `handleConnect` and `Unsubscribe`. However, there is no error returned to the caller to inform them that an error has occurred.

handleConnect
```
stream.OnConnect(stream.handlerConnect)

func (s *Stream) handlerConnect() {
	if s.PublicOnly {
		// errors are handled in the syncSubscriptions, so they are skipped here.
		_ = s.syncSubscriptions(WsEventSubscribe)
	} else {
		log.Error("*** PRIVATE API NOT IMPLEMENTED ***")
	}
}
```


Unsubscribe
```
func (s *Stream) Unsubscribe() {
	// errors are handled in the syncSubscriptions, so they are skipped here.
	_ = s.syncSubscriptions(WsEventUnsubscribe)
	s.Resubscribe(func(old []types.Subscription) (new []types.Subscription, err error) {
		// clear the subscriptions
		return []types.Subscription{}, nil
	})
}
```

Reply:
👍 got it, thanks